### PR TITLE
step_1_1

### PR DESCRIPTION
**Pull request by olgZZZ**
---
name: step_1_1
about: 
---
- Create a `Point` type which represents a 2D point (`x` and `y` coordinates). This type has to be `Copy` and `Default`.
- Create a `Polyline` type which represents a non-empty set of `Point`s of unknown size. This type has to be `Clone` and non-`Default`.

### CODE REVIEW

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 1]
```diff
@@ -1,4 +1,27 @@
...
+{
+  x : i32,
+  y : i32,
+}
+
+#[ allow( dead_code ) ]
+#[ derive( Clone, Copy, Debug ) ]
+struct Polyline
+{
+  point : Point,
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:13]: > `Polyline` type which represents a non-empty set of Points of unknown size

Now, the `Polyline` holds single value.
Please, update type.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 2]
```diff
@@ -1,4 +1,27 @@
...
+}
+
+
 fn main()
 {
-  println!( "Implement me!" );
+  let p = Point{ x : 47, y : 15 };
+  let p2 = p;
+  println!( "{:?}", p2 );
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:21]: Please, write proper tests for the copy and clone behavior.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 3]
```diff
@@ -1,4 +1,53 @@
...
+{
+  x : i32,
+  y : i32,
+}
+
+#[ allow( dead_code ) ]
+#[ derive( Clone, Debug, PartialEq ) ]
+struct Polyline
+{
+  points : Vec< Point >,
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:13]: > `Polyline` type which represents a non-empty set of Points of unknown size

To satisfy first requirement you need to implement constructor for the type. If the type visibility will be restricted by the module user will create an instance with only constructor.

Also, test the requirement : `of unknown size`

Comment (olgZZZ) [reply 1_concepts/1_1_default_clone_copy/src/main.rs:13]: Ok! I'lll do it today!

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 4]
```diff
@@ -1,4 +1,63 @@
...
+}
+
+impl Polyline
+{
+  pub fn new( point: Point ) -> Self
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:18]: Polyline should be a set of poits of unknown size. 

Please, update realization.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 5]
```diff
@@ -1,4 +1,63 @@
...
+
 fn main()
 {
-  println!( "Implement me!" );
+  let p = Point{ x : 47, y : 15 };
+  let p2 = p;
+  println!( "{:?}", p2 );
+
+  let p = Default::default();
+  let pol = Polyline{ points : vec![ p] };
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:35]: Please, take care of formatting. 

Also, do not duplicate tests, please.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 6]
```diff
@@ -8,14 +8,14 @@ struct Point
...
 {
-  points : Vec< Point >,
+  points : Vec< Box< T > >,
 }
 
-impl Polyline
+impl Polyline< Point >
 {
-  pub fn new( point: Point ) -> Self
+  pub fn new( point: Box< Point > ) -> Self
```
Comment (Barsik-sus) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:18]: You could implement `try_from( points : Vec< Point > ) -> Result< Self, SomeError >` which checks the vector IS empty or NOT, and if not empty — create Self

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 7]
```diff
@@ -1,4 +1,119 @@
...
+    assert_eq!( point1, point2 );
+  }
+
+  #[ test ]
+  fn polyline_clone_test()
+  {
+    let point = Point::default();
+    let polyline = Polyline::new( Box::new( point ) );
+    let polyline2 = polyline.clone();
+    assert_eq!( polyline, polyline2 );
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:99]: This test does not test the `Clone` behavior. It tests the content of instances. 
Please, extend test.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 8]
```diff
@@ -1,4 +1,119 @@
...
+}
+
+#[ derive( Debug, PartialEq ) ]
+struct EmptyPolyline;
+
+#[ allow( dead_code ) ]
+#[ derive( Clone, Debug, PartialEq ) ]
+struct Polyline< T: ?Sized >
+{
+  points : Vec< Box< T > >,
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:16]: Please, remove `Box` wrapper. It is not needed in the realization.

File: 1_concepts/1_1_default_clone_copy/src/main.rs [hunk 9]
```diff
@@ -1,4 +1,123 @@
...
+  #[ test ]
+  fn polyline_clone_test()
+  {
+    let point = Point::default();
+    let point2 = Point{ x : 1, y : 2 };
+    let polyline = Polyline::new( point );
+    let mut polyline2 = polyline.clone();
+    assert_eq!( polyline, polyline2 );
+
+    polyline2.points = vec![ point2 ];
```
Comment (dmvict) [inline 1_concepts/1_1_default_clone_copy/src/main.rs:102]: It is better to use public interfaces. The field can be hidden for the routine.

**Comment by dmvict** [comment]
Can be merged.

**Pull request merged**